File: src/restore-dest.ts
import * as p from "@clack/prompts";
import type { Config, DatabaseItem } from "./config.ts";
import { dbKey, configImage, getParentItem } from "./config.ts";
import {
  createDatabase,
  databaseExists,
  dropDatabase,
  ensureDatabaseLogin,
  verifyConnection,
  type ResolvedDb,
} from "./docker.ts";
import { getDbPassword, setDbPassword, type Session } from "./metadata.ts";
import {
  confirmOrYes,
  connectWithRetry,
  promptConfirmedPassword,
  resolveDbPassword,
  selectNestedCreatePassword,
  selectNestedRestoreAction,
} from "./prompt.ts";

export type PrepareRestoreDestinationOpts = {
  config: Config;
  session: Session | null;
  destItem: DatabaseItem;
  image: string;
  yes?: boolean;
  /** Prompt for nested restore; flat destinations use confirmOrYes with this message. */
  confirmMessage: string;
};

export type PrepareRestoreDestinationResult =
  | { cancelled: true }
  | { destDb: ResolvedDb; intoExisting: boolean };

async function resolveConnectedDb(opts: {
  config: Config;
  session: Session | null;
  item: DatabaseItem;
  role: "source" | "destination";
  maintenance?: boolean;
}): Promise<ResolvedDb> {
  const image = configImage(opts.config);
  const key = dbKey(opts.item.key);

  const password = await connectWithRetry({
    label: opts.item.key,
    getPassword: () =>
      resolveDbPassword({
        session: opts.session,
        rememberPassword: opts.config.rememberPassword,
        item: opts.item,
      }),
    setPassword: async (pw) => {
      if (opts.config.rememberPassword && opts.session) {
        await setDbPassword(opts.session, key, pw);
      }
    },
    connect: async (pw) => {
      const db: ResolvedDb = { ...opts.item, password: pw };
      await verifyConnection(
        image,
        opts.role,
        opts.item.key,
        db,
        opts.maintenance ? { database: "postgres" } : undefined,
      );
    },
  });

  return { ...opts.item, password };
}

async function ensureDestDatabase(
  config: Config,
  dest: ResolvedDb,
  destName: string,
  yes?: boolean,
): Promise<{ created: boolean }> {
  const image = configImage(config);
  const exists = await databaseExists(image, dest);
  if (exists) return { created: false };
  const create = await confirmOrYes(
    `Database "${dest.database}" does not exist on destination (${destName}). Create it?`,
    yes,
    false,
  );
  if (!create) {
    throw new Error("Destination database does not exist. Aborted.");
  }
  p.log.step(`Creating database "${dest.database}"…`);
  await createDatabase(image, dest);
  p.log.success(`Created "${dest.database}"`);
  return { created: true };
}

/** Shared flat + nested destination prep for restore and dump-restore flows. */
export async function prepareRestoreDestination(
  opts: PrepareRestoreDestinationOpts,
): Promise<PrepareRestoreDestinationResult> {
  const { config, session, destItem, image, yes, confirmMessage } = opts;

  if (destItem.nested && destItem.parentKey) {
    const parentItem = getParentItem(config, destItem.parentKey);
    if (!parentItem) {
      throw new Error(
        `Nested destination "${destItem.key}" needs parent "${destItem.parentKey}" with user+database for connection verify.`,
      );
    }

    p.log.step(`Verifying parent "${parentItem.key}"…`);
    const parentDb = await resolveConnectedDb({
      config,
      session,
      item: parentItem,
      role: "destination",
    });
    p.log.success(`Parent "${parentItem.key}" OK`);

    const parentLogin = parentDb.database;
    const childTarget = { ...parentDb, database: destItem.database };
    const connectOpts = { connectDatabase: parentLogin };

    const childExists = await databaseExists(image, childTarget, connectOpts);

    const action = await selectNestedRestoreAction(confirmMessage, childExists);
    if (action === "no") {
      return { cancelled: true };
    }

    if (action === "drop") {
      p.log.step(`Dropping database "${destItem.database}"…`);
      await dropDatabase(image, childTarget, connectOpts);
      p.log.step(`Creating database "${destItem.database}"…`);
      await createDatabase(image, childTarget, connectOpts);
      p.log.success(`Recreated "${destItem.database}"`);
    } else if (action === "create") {
      p.log.step(`Creating database "${destItem.database}"…`);
      await createDatabase(image, childTarget, connectOpts);
      p.log.success(`Created "${destItem.database}"`);
    }

    if (action === "create" || action === "drop") {
      if (parentDb.user === destItem.user) {
        return {
          destDb: { ...destItem, user: parentDb.user, password: parentDb.password },
          intoExisting: false,
        };
      }

      const childKey = dbKey(destItem.key);
      const saved =
        session && config.rememberPassword
          ? await getDbPassword(session, childKey)
          : null;
      const pwSource = yes
        ? "parent"
        : await selectNestedCreatePassword({ hasSaved: Boolean(saved) });

      if (pwSource === "parent") {
        return {
          destDb: { ...destItem, user: parentDb.user, password: parentDb.password },
          intoExisting: false,
        };
      }
      if (pwSource === "saved") {
        if (!saved) throw new Error(`No saved password for "${destItem.key}"`);
        return {
          destDb: { ...destItem, password: saved },
          intoExisting: false,
        };
      }

      const password = await promptConfirmedPassword(
        `password for ${destItem.key} (${destItem.user})`,
      );
      p.log.step(`Ensuring login "${destItem.user}"…`);
      await ensureDatabaseLogin(image, parentDb, {
        user: destItem.user,
        password,
        database: destItem.database,
        connectDatabase: parentLogin,
      });
      if (config.rememberPassword && session) {
        await setDbPassword(session, childKey, password);
      }
      p.log.success(`Login "${destItem.user}" ready`);
      return {
        destDb: { ...destItem, password },
        intoExisting: false,
      };
    }

    return {
      destDb: { ...destItem, user: parentDb.user, password: parentDb.password },
      intoExisting: true,
    };
  }

  const destDb = await resolveConnectedDb({
    config,
    session,
    item: destItem,
    role: "destination",
    maintenance: true,
  });
  const { created } = await ensureDestDatabase(config, destDb, destItem.key, yes);

  const confirmed = await confirmOrYes(confirmMessage, yes, false);
  if (!confirmed) {
    return { cancelled: true };
  }

  return { destDb, intoExisting: !created };
}

File: src/internal/app/restore_dest.go
package app

import (
	"fmt"

	"github.com/md-redwan-hossain/dumpmgr/src/internal/config"
	"github.com/md-redwan-hossain/dumpmgr/src/internal/docker"
	"github.com/md-redwan-hossain/dumpmgr/src/internal/metadata"
	"github.com/md-redwan-hossain/dumpmgr/src/internal/prompt"
)

type PrepareRestoreDestinationOpts struct {
	Config         *config.Config
	Session        *metadata.Session
	DestItem       config.DatabaseItem
	Image          string
	Yes            bool
	ConfirmMessage string
}

type PrepareRestoreDestinationResult struct {
	Cancelled    bool
	DestDB       docker.ResolvedDB
	IntoExisting bool
}

// PrepareRestoreDestination handles flat and nested destination setup for restore and dump-restore.
func PrepareRestoreDestination(opts PrepareRestoreDestinationOpts) (PrepareRestoreDestinationResult, error) {
	cfg := opts.Config
	session := opts.Session
	destItem := opts.DestItem
	image := opts.Image
	yes := opts.Yes
	confirmMessage := opts.ConfirmMessage

	if destItem.Nested && destItem.ParentKey != "" {
		parentItem := config.GetParentItem(cfg, destItem.ParentKey)
		if parentItem == nil {
			return PrepareRestoreDestinationResult{}, fmt.Errorf(
				`nested destination %q needs parent %q with user+database for connection verify`,
				destItem.Key, destItem.ParentKey,
			)
		}

		prompt.LogStep(fmt.Sprintf("Verifying parent %q…", parentItem.Key))
		parentDB, err := resolveConnectedDB(cfg, session, *parentItem, "destination", false)
		if err != nil {
			return PrepareRestoreDestinationResult{}, err
		}
		prompt.LogSuccess(fmt.Sprintf("Parent %q OK", parentItem.Key))

		parentLogin := parentDB.Database
		childTarget := docker.ResolvedDB{
			DatabaseItem: config.DatabaseItem{
				Key: destItem.Key, Host: parentDB.Host, Port: parentDB.Port,
				User: parentDB.User, Database: destItem.Database, Nested: true, ParentKey: destItem.ParentKey,
			},
			Password: parentDB.Password,
		}
		childExists, err := docker.DatabaseExists(image, childTarget, parentLogin)
		if err != nil {
			return PrepareRestoreDestinationResult{}, err
		}

		action, err := prompt.SelectNestedRestoreAction(confirmMessage, childExists)
		if err != nil {
			return PrepareRestoreDestinationResult{}, err
		}
		if action == prompt.NestedNo {
			return PrepareRestoreDestinationResult{Cancelled: true}, nil
		}

		if action == prompt.NestedDrop {
			prompt.LogStep(fmt.Sprintf("Dropping database %q…", destItem.Database))
			if err := docker.DropDatabase(image, childTarget, parentLogin); err != nil {
				return PrepareRestoreDestinationResult{}, err
			}
			prompt.LogStep(fmt.Sprintf("Creating database %q…", destItem.Database))
			if err := docker.CreateDatabase(image, childTarget, parentLogin); err != nil {
				return PrepareRestoreDestinationResult{}, err
			}
			prompt.LogSuccess(fmt.Sprintf("Recreated %q", destItem.Database))
		} else if action == prompt.NestedCreate {
			prompt.LogStep(fmt.Sprintf("Creating database %q…", destItem.Database))
			if err := docker.CreateDatabase(image, childTarget, parentLogin); err != nil {
				return PrepareRestoreDestinationResult{}, err
			}
			prompt.LogSuccess(fmt.Sprintf("Created %q", destItem.Database))
		}

		if action == prompt.NestedCreate || action == prompt.NestedDrop {
			if parentDB.User == destItem.User {
				item := destItem
				item.User = parentDB.User
				return PrepareRestoreDestinationResult{
					DestDB:       docker.ResolvedDB{DatabaseItem: item, Password: parentDB.Password},
					IntoExisting: false,
				}, nil
			}

			childKey := config.DBKey(destItem.Key)
			var saved string
			if session != nil && cfg.RememberPassword {
				saved, _ = metadata.GetDBPassword(session, childKey)
			}
			var pwSource prompt.NestedCreatePasswordSource
			if yes {
				pwSource = prompt.PasswordParent
			} else {
				pwSource, err = prompt.SelectNestedCreatePassword(saved != "")
				if err != nil {
					return PrepareRestoreDestinationResult{}, err
				}
			}

			switch pwSource {
			case prompt.PasswordParent:
				item := destItem
				item.User = parentDB.User
				return PrepareRestoreDestinationResult{
					DestDB:       docker.ResolvedDB{DatabaseItem: item, Password: parentDB.Password},
					IntoExisting: false,
				}, nil
			case prompt.PasswordSaved:
				if saved == "" {
					return PrepareRestoreDestinationResult{}, fmt.Errorf("no saved password for %q", destItem.Key)
				}
				return PrepareRestoreDestinationResult{
					DestDB:       docker.ResolvedDB{DatabaseItem: destItem, Password: saved},
					IntoExisting: false,
				}, nil
			case prompt.PasswordNew:
				password, err := prompt.ConfirmedPassword(fmt.Sprintf("password for %s (%s)", destItem.Key, destItem.User))
				if err != nil {
					return PrepareRestoreDestinationResult{}, err
				}
				prompt.LogStep(fmt.Sprintf("Ensuring login %q…", destItem.User))
				if err := docker.EnsureDatabaseLogin(image, parentDB, docker.EnsureLoginOpts{
					User: destItem.User, Password: password, Database: destItem.Database, ConnectDatabase: parentLogin,
				}); err != nil {
					return PrepareRestoreDestinationResult{}, err
				}
				if cfg.RememberPassword && session != nil {
					if err := metadata.SetDBPassword(session, childKey, password); err != nil {
						return PrepareRestoreDestinationResult{}, err
					}
				}
				prompt.LogSuccess(fmt.Sprintf("Login %q ready", destItem.User))
				return PrepareRestoreDestinationResult{
					DestDB:       docker.ResolvedDB{DatabaseItem: destItem, Password: password},
					IntoExisting: false,
				}, nil
			}
		}

		item := destItem
		item.User = parentDB.User
		return PrepareRestoreDestinationResult{
			DestDB:       docker.ResolvedDB{DatabaseItem: item, Password: parentDB.Password},
			IntoExisting: true,
		}, nil
	}

	destDB, err := resolveConnectedDB(cfg, session, destItem, "destination", true)
	if err != nil {
		return PrepareRestoreDestinationResult{}, err
	}
	created, err := ensureDestDatabase(cfg, destDB, destItem.Key, yes)
	if err != nil {
		return PrepareRestoreDestinationResult{}, err
	}

	confirmed, err := prompt.ConfirmOrYes(confirmMessage, yes, false)
	if err != nil {
		return PrepareRestoreDestinationResult{}, err
	}
	if !confirmed {
		return PrepareRestoreDestinationResult{Cancelled: true}, nil
	}

	return PrepareRestoreDestinationResult{
		DestDB:       destDB,
		IntoExisting: !created,
	}, nil
}
